### PR TITLE
Add packing for int8 1x1 convolution and support the int8 group convolution on X86

### PR DESCRIPTION
**Pull request by llyfacebook**
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/dmlc/tvm/blob/master/CONTRIBUTORS.md#reviewers).

### CODE REVIEW

**Comment by llyfacebook** [comment]
cc @yzhliu @yidawang @anijain2305 
Basically, we want to upstream some changes, which are used in our optimization for quantized ResNext101.

**Comment by tqchen** [comment]
@llyfacebook Please check the testcase, because of not all test machines support avx512, we need to optionally skip them.

**Comment by tqchen** [comment]
@yzhliu @kevinthesun can you help follow up and review this PR?

**Comment by llyfacebook** [comment]
Thanks, I will skip the tests.

File: topi/python/topi/nn/conv2d.py [hunk 1]
```diff
@@ -95,19 +95,33 @@ def conv2d_alter_layout(attrs, inputs, tinfos, F):
...
+        _, CI, IH, IW = [x.value for x in data.shape]
+    elif data_layout == 'NHWC':
+        _, IH, IW, CI = [x.value for x in data.shape]
+    elif data_layout == 'HWCN':
+        IH, IW, CI, _ = [x.value for x in data.shape]
+    else:
+        raise ValueError("not support this layout {} yet".format(data_layout))
+
+
+    if data_layout == 'NHWC':
```
Comment (yzhliu) [inline topi/python/topi/nn/conv2d.py]: shall we use `kernel_layout` instead? as `data_layout` might not be necessarily binded to `kernal_layout`.
I'm actually a bit confused with the int8 conv layout, for NHWC data, what kernal layout is it?

File: topi/python/topi/x86/conv2d.py [hunk 2]
```diff
@@ -62,6 +63,9 @@ def _create_tuning_space(cfg, data, kernel, strides, padding, dilation, layout):
     if layout == 'NCHW':
         n, ic, h, w = dshape
         oc, _, kh, kw = kshape
+    elif layout == 'NHWC':
+        n, h, w, ic = dshape
+        oc, _, kh, kw = kshape
```
Comment (yzhliu) [inline topi/python/topi/x86/conv2d.py]: `kh, kw, oc, _` ?

File: topi/python/topi/x86/conv2d.py [hunk 3]
```diff
@@ -93,12 +97,14 @@ def _create_tuning_space(cfg, data, kernel, strides, padding, dilation, layout):
...
 
-@autotvm.register_topi_compute(conv2d, 'cpu', 'direct')
+@autotvm.register_topi_compute(conv2d, 'cpu', ['direct'])
 def _declaration_conv(cfg, data, kernel, strides, padding, dilation, layout, out_dtype):
     out_dtype = data.dtype if out_dtype is None else out_dtype
     padding = padding if isinstance(padding, (tuple, list)) else (padding, padding)
     strides = strides if isinstance(strides, (tuple, list)) else (strides, strides)
     dilation = dilation if isinstance(dilation, (tuple, list)) else (dilation, dilation)
+
+    _, _, kh, kw = get_const_tuple(kernel.shape)
```
Comment (yzhliu) [inline topi/python/topi/x86/conv2d.py]: so, for data=NHWC & fp32, kernel=HWIO, while for data=NHWC & int8, kernel=OIHW?

File: topi/python/topi/x86/conv2d.py [hunk 4]
```diff
@@ -226,6 +238,58 @@ def traverse(op):
...
+
+            args = [s, cfg, data_vec, conv_out, outs[0]]
+            if data.dtype == 'uint8':
+                # int8 conv kernel is 7-dim
+                kh, kw, _, _, _ = get_const_tuple(kernel.shape)
+                if kh == 1 and kw == 1:
+                    conv2d_avx_1x1._schedule_conv_nhwc_pack_int8(*args)
+                else:
+                    raise ValueError("Only support 1x1 kernel with "
+                                     "schedule template.")
```
Comment (yzhliu) [inline topi/python/topi/x86/conv2d.py]: please make the fatal msg more detailed other than just "schedule template"

**Comment by tqchen** [comment]
@llyfacebook please update as per review comments

File: topi/python/topi/nn/conv2d.py (hunk 1 above)
Comment (llyfacebook) [reply topi/python/topi/nn/conv2d.py]: I think I was mainly following the data layout and kernel layout corresponding relationship  here: https://github.com/dmlc/tvm/blob/147ea3b0ca147b527086228d524a2f68f872112d/topi/python/topi/nn/conv2d.py#L284

File: topi/python/topi/x86/conv2d.py (hunk 3 above)
Comment (llyfacebook) [reply topi/python/topi/x86/conv2d.py]: Thanks. I will spend some time this week unifying them.

**Comment by tqchen** [comment]
@yzhliu please followup on this PR

File: topi/python/topi/x86/conv2d.py [hunk 5]
```diff
@@ -93,21 +97,30 @@ def _create_tuning_space(cfg, data, kernel, strides, padding, dilation, layout):
...
     strides = strides if isinstance(strides, (tuple, list)) else (strides, strides)
     dilation = dilation if isinstance(dilation, (tuple, list)) else (dilation, dilation)
+
     if layout == 'NCHW':
         _create_tuning_space(cfg, data, kernel, strides, padding, dilation, layout)
         if cfg.is_fallback:
             _get_default_config(cfg, data, kernel, strides, padding, out_dtype)
         return _declaration_conv_impl(cfg, data, kernel, strides,
                                       padding, dilation, layout, out_dtype)
+    # KHOI kernel layout is for NHWC and HWCN
```
Comment (yzhliu) [inline topi/python/topi/x86/conv2d.py]:
```suggestion
    # HWOI kernel layout is for NHWC and HWCN
```

**Comment by yzhliu** [comment]
Thanks @llyfacebook This is merged

**Pull request merged**